ethereum/rpc/types/block.go

```diff
@@ -103,3 +106,63 @@ func (bn BlockNumber) TmHeight() *int64 {
 	height := bn.Int64()
 	return &height
 }
+
+// BlockNumberOrHash represents a block number or a block hash.
+type BlockNumberOrHash struct {
+	BlockNumber *BlockNumber `json:"blockNumber,omitempty"`
+	BlockHash   *common.Hash `json:"blockHash,omitempty"`
+}
+
+func (bnh *BlockNumberOrHash) UnmarshalJSON(data []byte) error {
+	type erased BlockNumberOrHash
+	e := erased{}
+	err := json.Unmarshal(data, &e)
+	if err == nil {
+		if e.BlockNumber != nil && e.BlockHash != nil {
+			return fmt.Errorf("cannot specify both BlockHash and BlockNumber, choose one or the other")
+		}
+		bnh.BlockNumber = e.BlockNumber
+		bnh.BlockHash = e.BlockHash
+		return nil
+	}
```

Comment:
this can be grouped in one private function that checks for the unmarshaling of the full struct

Reply:
fixed